## DATA MANIPULATION WITH PANDAS
#### Page Visits Funnel
A online retailer would like to analyze data on visits to their website. In this notebook I build a **sales funnel**, which is a description of how many people continue to the next step of a multi-step process. 


In this case, our funnel is going to describe the following process:

1. A user visits the website
2. A user adds a product to their cart
3. A user clicks “checkout”
4. A user actually purchases a product

1. Inspect the DataFrames using print and head:

* visits lists all of the users who have visited the website
* cart lists all of the users who have added a t-shirt to their cart
* checkout lists all of the users who have started the checkout
* purchase lists all of the users who have purchased a t-shirt

In [1]:
import pandas as pd
import numpy as np
import datetime as dt

visits = pd.read_csv(r'E:\Jupyter_Projects\Multiple Tables Page Visits Funnel\visits.csv')
cart = pd.read_csv(r'E:\Jupyter_Projects\Multiple Tables Page Visits Funnel\cart.csv')
checkout = pd.read_csv(r'E:\Jupyter_Projects\Multiple Tables Page Visits Funnel\checkout.csv')
purchase = pd.read_csv(r'E:\Jupyter_Projects\Multiple Tables Page Visits Funnel\purchase.csv')


In [2]:
print(visits.head())
print(cart.head())
print(checkout.head())
print(purchase.head())

#visits_no_duplicate_ids 
visits = visits.drop_duplicates(subset='user_id', keep='first').reset_index(drop=True)
#cart_no_duplicate_ids 
cart = cart.drop_duplicates(subset='user_id', keep='first').reset_index(drop=True)
#checkout_no_duplicate_ids 
checkout = checkout.drop_duplicates(subset='user_id', keep='first').reset_index(drop=True)
#purchase_no_duplicate_ids 
purchase = purchase.drop_duplicates(subset='user_id', keep='first').reset_index(drop=True)

                                user_id           visit_time
0  943647ef-3682-4750-a2e1-918ba6f16188  2017-04-07 15:14:00
1  0c3a3dd0-fb64-4eac-bf84-ba069ce409f2  2017-01-26 14:24:00
2  6e0b2d60-4027-4d9a-babd-0e7d40859fb1  2017-08-20 08:23:00
3  6879527e-c5a6-4d14-b2da-50b85212b0ab  2017-11-04 18:15:00
4  a84327ff-5daa-4ba1-b789-d5b4caf81e96  2017-02-27 11:25:00
                                user_id            cart_time
0  2be90e7c-9cca-44e0-bcc5-124b945ff168  2017-11-07 20:45:00
1  4397f73f-1da3-4ab3-91af-762792e25973  2017-05-27 01:35:00
2  a9db3d4b-0a0a-4398-a55a-ebb2c7adf663  2017-03-04 10:38:00
3  b594862a-36c5-47d5-b818-6e9512b939b3  2017-09-27 08:22:00
4  a68a16e2-94f0-4ce8-8ce3-784af0bbb974  2017-07-26 15:48:00
                                user_id        checkout_time
0  d33bdc47-4afa-45bc-b4e4-dbe948e34c0d  2017-06-25 09:29:00
1  4ac186f0-9954-4fea-8a27-c081e428e34e  2017-04-07 20:11:00
2  3c9c78a7-124a-4b77-8d2e-e1926e011e7d  2017-07-13 11:38:00
3  89fe330a-8966-4756-8f

2. Combine visits and cart using a left merge.
3.  How long is the merged DataFrame?

In [3]:
visit_to_cart = visits.merge(cart, how="left")
visit_to_cart_rows = len(visit_to_cart)
print(f'The amount of visits on the page was {visit_to_cart_rows}.')

The amount of visits on the page was 2000.


4. How many of the timestamps are null for the column cart_time?

    What do these null rows mean?

In [4]:
null_cart_times = len(visit_to_cart[visit_to_cart.cart_time.isnull()])
num_added_to_cart = visit_to_cart_rows-null_cart_times
print(f'The amount of vistors that did not add anything to their cart was {null_cart_times}.')
print(f'The amount of visitors that did add a product to their cart was {num_added_to_cart}.')

The amount of vistors that did not add anything to their cart was 1652.
The amount of visitors that did add a product to their cart was 348.


5. What percent of users who visited Cool T-Shirts Inc. ended up not placing a t-shirt in their cart?


In [5]:
print(f'The percentage of vistors that did not add products to their cart: \n{float(null_cart_times)/(visit_to_cart_rows)}.')
print('___________________________________')
print(f'The percentage of visitors who did add products to their cart:\n{float(num_added_to_cart)/(visit_to_cart_rows)}.')

The percentage of vistors that did not add products to their cart: 
0.826.
___________________________________
The percentage of visitors who did add products to their cart:
0.174.


6.  Repeat the left merge for cart and checkout and count null values. What percentage of users put items in their cart, but did not proceed to checkout?

In [6]:
cart_to_checkout = cart.merge(checkout, how='left')
cart_to_checkout_rows = len(cart_to_checkout)

null_checkout_times = len(cart_to_checkout[cart_to_checkout.checkout_time.isnull()])
num_proceed_to_checkout = cart_to_checkout_rows-null_checkout_times
print(f'The amount of visitors who did not proceed to checkout after they added an item to their cart was {null_checkout_times}.')
print(f'The amount of visitors who proceeded to checkout after they added an item to their cart was {num_proceed_to_checkout}.')
print('___________________________________')
print(f'The percentage of visitors that did not proceed to checkout after adding a product to their cart was:\n{round(float(null_checkout_times)/(cart_to_checkout_rows),3)}.')
print(f'The percentage of visitors that did proceed to checkout after adding a product to their cart was:\n{round(float(num_proceed_to_checkout)/(cart_to_checkout_rows),3)}.')

The amount of visitors who did not proceed to checkout after they added an item to their cart was 122.
The amount of visitors who proceeded to checkout after they added an item to their cart was 226.
___________________________________
The percentage of visitors that did not proceed to checkout after adding a product to their cart was:
0.351.
The percentage of visitors that did proceed to checkout after adding a product to their cart was:
0.649.


8.  What percentage of users proceeded to checkout, but did not purchase a t-shirt?

In [7]:
checkout_to_purchase = checkout.merge(purchase, how="left")
checkout_to_purchase_rows = len(checkout_to_purchase)

null_purchase_times = len(checkout_to_purchase[checkout_to_purchase.purchase_time.isnull()])
num_purchased = checkout_to_purchase_rows-null_purchase_times
print(f'The number of visitors who did not purchase once they proceeded to checkout was {null_purchase_times}.')
print(f'The number of visitors who did purchase after proceededing to checkout was {num_purchased}.')
print('___________________________________')
print(f'The percentage of visitors that did not purchase after proceeding to checkout was:\n {round(float(null_purchase_times)/checkout_to_purchase_rows,3)}.')
print(f'The percentage of visitors that purchased a product after proceeding to checkout was:\n {round(float(num_purchased)/checkout_to_purchase_rows,3)}.')


The number of visitors who did not purchase once they proceeded to checkout was 82.
The number of visitors who did purchase after proceededing to checkout was 144.
___________________________________
The percentage of visitors that did not purchase after proceeding to checkout was:
 0.363.
The percentage of visitors that purchased a product after proceeding to checkout was:
 0.637.


7. Merge all four steps of the funnel, in order, using a series of left merges. Save the results to the variable all_data.

    Examine the result using print and head.

In [8]:
print(f'The percentage of vistors that did not add products to their cart: \n{float(null_cart_times)/(visit_to_cart_rows)}')
print('___________________________________')
print(f'The percentage of visitors that did not proceed to checkout after adding a product to their cart was:\n{round(float(null_checkout_times)/(cart_to_checkout_rows),3)}.')
print('___________________________________')
print(f'The percentage of visitors that did not purchase after proceeding to checkout was:\n{round(float(null_purchase_times)/checkout_to_purchase_rows,3)}.')

The percentage of vistors that did not add products to their cart: 
0.826
___________________________________
The percentage of visitors that did not proceed to checkout after adding a product to their cart was:
0.351.
___________________________________
The percentage of visitors that did not purchase after proceeding to checkout was:
0.363.


The step in the funnel that appears to have the greatest amount of loss is between "visitor" and "cart". 82.6% of site visitors leave the site before adding a product to the cart. The company may improve their conversion rate between these two steps by reducing friction of doing so, offering a first purchase discount or some other call to action (CTA). They could initiate a A/B campaign in multiple markets that allows the organization to obtain and analyze data on whether A or B results in greater results. 

In [10]:
visits.head()

user_id           visit_time
0  943647ef-3682-4750-a2e1-918ba6f16188  2017-04-07 15:14:00
1  0c3a3dd0-fb64-4eac-bf84-ba069ce409f2  2017-01-26 14:24:00
2  6e0b2d60-4027-4d9a-babd-0e7d40859fb1  2017-08-20 08:23:00
3  6879527e-c5a6-4d14-b2da-50b85212b0ab  2017-11-04 18:15:00
4  a84327ff-5daa-4ba1-b789-d5b4caf81e96  2017-02-27 11:25:00

## Converting visit_time and purchase_time format from 'str' to 'datetime'

### data cleaning

In [11]:
visits['visit_time'] = pd.to_datetime(visits['visit_time'], format='%Y-%m-%d %H:%M:%S')
purchase['purchase_time'] = pd.to_datetime(purchase['purchase_time'], format='%Y-%m-%d %H:%M:%S')
# df['dates'] = pd.to_datetime(df['dates'], format='%Y%m%d')

Merging all dataframes into all_data and subtracting visit time from purchase time, adding result to all_data

In [15]:
all_data = visits.merge(cart, how="left").merge(checkout, how="left").merge(purchase, how="left")

all_data['time_to_purchase'] = all_data['purchase_time'] - all_data['visit_time']
all_data.dtypes

user_id                      object
visit_time           datetime64[ns]
cart_time                    object
checkout_time                object
purchase_time        datetime64[ns]
time_to_purchase    timedelta64[ns]
dtype: object

Inspecting time_to_purchase column

In [13]:
print(all_data.time_to_purchase)

0                  NaT
1      0 days 00:44:00
2                  NaT
3                  NaT
4                  NaT
             ...      
1995               NaT
1996               NaT
1997               NaT
1998               NaT
1999               NaT
Name: time_to_purchase, Length: 2000, dtype: timedelta64[ns]


Calculating the average purchase time - from visit to purchase

In [14]:
print(all_data.time_to_purchase.mean())

0 days 00:42:33.333333333
